src/routes/preHandlers/courses.js

```diff
@@ -202,13 +214,20 @@ exports.authorizeCourseEdit = async (req) => {
       if (courseBillsWithoutCreditNote.length > req.payload.expectedBillsCount) throw Boom.conflict();
     }
 
-    await this.checkInterlocutors(req, companies[0]);
+    await this.checkInterlocutors(req, course);
 
     if (get(req, 'payload.contact')) {
-      const isCompanyRepContactAndUpdated = !!get(req, 'payload.companyRepresentative') &&
+      const isCompanyRepContact = !!get(req, 'payload.companyRepresentative') &&
         UtilsHelper.areObjectIdsEquals(course.companyRepresentative, get(course, 'contact._id'));
-      const hasUserAccessToCourseCompany = companies.some(c => UtilsHelper.hasUserAccessToCompany(credentials, c));
-      if (!isRofOrAdmin && !(isCompanyRepContactAndUpdated && hasUserAccessToCourseCompany)) throw Boom.forbidden();
+      let canUserUpdateCompanyRep = false;
+      if (course.type === INTRA) {
+        canUserUpdateCompanyRep = companies.some(c => UtilsHelper.hasUserAccessToCompany(credentials, c));
+      }
+      if (course.type === INTRA_HOLDING) {
+        canUserUpdateCompanyRep = UtilsHelper.areObjectIdsEquals(get(credentials, 'holding._id'), course.holding) &&
+          [HOLDING_ADMIN].includes(userHoldingRole);
+      }
+      if (!isRofOrAdmin && !(isCompanyRepContact && canUserUpdateCompanyRep)) throw Boom.forbidden();
```

Comment:
Que penses-tu de definir la constante `isCompanyRepContact` juste au dessus de ce if qui est l'endroit ou on l'utilise ? Je trouve que ca clarifie un peu

Reply:
+ j'ai l'impression qu'il manque des tests sur la partie contact , notamment sur la condition `if (!isRofOrAdmin && !(isCompanyRepContact && canUserUpdateCompanyRep)) throw Boom.forbidden();`

Par ex: 
- ETQ holding admin j'edite le companyRepresentative qui n'est pas contact mais j'envoies tout de meme contact dans la payload => 403
- ETQ holding admin , j'edite le companyRepresentative qui est contact mais ma holding est differente de celle de la formation => 403

Reply:
1: test linge 2273 'should return 403 if holding admin try to update contact and company rep which is not contact'
2 : je me demande si on a besoin de cette variable canUserUpdateCompanyRep, en fait c'est déjà vérifié dans checkAuthorization ça non ?

3: je me tâte à supprimer le test ligne 2059 je trouve qu'il fait doublon avec le test 2260

Reply:
1. ah yes merci ! 
2. hmm non je n'ai pas l'impression qu'il y ai de doublon. Tu voulais plutot dire dans checkInterlocutors non ? Je ne vois pas le doublon, la on veut savoir si le contact est aussi le chargé de formation structure pour ensuite savoir si le qqn qui a un role client a le droit d'editer le contact 

Mais par contre je pense qu'on peut essayer de reorganiser les conditions pour y voir plus clair. D'ailleurs la condition !isRofOrAdmin ne me parait pas assez inclusive: les formateurs n'ont pas le droit d'editer le contact. Je pense qu'on peut definir un premier block dans le `if(get(req, 'payload.contact'))`

```
if (userVendorRole === TRAINER) throw Boom.forbidden()
const isCompanyRepContact = !!get(req, 'payload.companyRepresentative') &&
        UtilsHelper.areObjectIdsEquals(course.companyRepresentative, get(course, 'contact._id'));
if (hasClientRole && !isCompanyRepContactAndUpdated )  throw Boom.forbidden();
```
Puis ensuite 
```
      let canUserUpdateCompanyRep = false;
      if (course.type === INTRA) {
        canUserUpdateCompanyRep = companies.some(c => UtilsHelper.hasUserAccessToCompany(credentials, c));
      }
      if (course.type === INTRA_HOLDING) {
        canUserUpdateCompanyRep = UtilsHelper
          .areObjectIdsEquals(get(credentials, 'holding._id'), course.holding);
      }
      if (hasClientRole && !canUserUpdateCompanyRep)) throw Boom.forbidden();
```
+ 
Comme les tests pour cette route s'etendent sur 600 lignes, que penses-tu d'essayer de les grouper par ex par type de formation puis pour chaque type par role? (ou l'inverse ou autre si tu trouves un meilleur groupement) J'ai l'impression qu'on y verrait plus clair